### ISSUE TITLE
FR: Tailscale SSH should allow "users": "autogroup:self"

### ISSUE THREAD

**Issue by skorokithakis** [tailscale/tailscale#5562]
### What are you trying to do?

Tailscale SSH with `autogroup:nonroot` currently lets you log in to any machine as any user (except root), so it lets you impersonate anyone. I'm trying to restrict access to people as only themselves.

### How should we solve this?

It would be good if there were an `autogroup:self` directive that only let you log in as yourself (perhaps your GitHub username?).

### What is the impact of not solving this?

Any user will be able to log in as any user, we haven't found a solution to this yet.

**Comment by mayakacz** [comment]
I have a few questions on implementation for functionality like this:
* Do you want each individual to have their own just-in-time created user on the host (e.g., "alice@tailscale.com"), or each individual (in an ACL) to log in as the same just-in-time created user, e.g., "devops"?
* If you want multiple individuals to log in to the same host user "devops", then what happens if there is already a session in process on the host for that user? Is another individual blocked from starting a session as that user? i.e. are the sessions exclusive?
* For users that are just-in-time provisioned, which groups on the host should they have? How should sudo permissions work for those users?
Thanks!

**Comment by skorokithakis** [comment]
I think that the cleanest thing to do would be to have people created just in time, if they don't exist, and assigned an additional `tailscale` group.

This lets me provision users beforehand if I want, or I can just trust that they'll be created with their GH username, and I can assign permissions to their group to allow access that way. It also preserves auditing, as I can tell who is who in logs.

Does that make sense?

**Comment by mayakacz** [comment]
> and assigned an additional tailscale group.

Ah, I meant the host groups. How do you want to assign host groups to the host users that are created just in time?

**Comment by skorokithakis** [comment]
Hm, what do you mean? Presumably, if Tailscale can create the user, it can also assign them to a pre-created `tailscale` group, no? Then all permissions (including sudo) will be governed by the group.

**Comment by mayakacz** [comment]
Does the tailscale host group already exist, or would you want it created at that point in time as well?
Assuming it's created at that point in time, how do you want to customize the host group?

**Comment by maisem** [comment]
> Presumably, if Tailscale can create the user, it can also assign them to a pre-created tailscale group, no?

Yes Tailscale can assign them to that group, the question here is whether you want to control which groups are assigned based on ACLs?

**Comment by skorokithakis** [comment]
@mayakacz I haven't given it much thought, but it seems that I can set up the group (or it can be set up automatically) when I install Tailscale, and then I can add the permissions I want to it. I think this would be the most convenient for people.

@maisem That's a very good question, I think that can be a separate feature later on. It seems to me that you'd always want a general "tailscale" group, and later on you could add the option to control the group assignments via ACLs? It does sound like a very useful feature, though.

**Comment by maisem** [comment]
@skorokithakis I worry about a general group named `tailscale` as that might conflict with some installations of tailscale. It would be safer IMO if it was user defined.

**Comment by skorokithakis** [comment]
I agree, I assume it would be a cli command when logging in to the tailnet, or similar.

**Comment by gebrk** [comment]
Responding to @mayakacz's original questions - I'm assuming we're talking about `autogroup:self` being allowed for the `users`  key in an ssh ACL:

> * Do you want each individual to have their own just-in-time created user on the host (e.g., "[alice@tailscale.com](mailto:alice@tailscale.com)"), or each individual (in an ACL) to log in as the same just-in-time created user, e.g., "devops"
> * For users that are just-in-time provisioned, which groups on the host should they have? How should sudo permissions work for those users?

I think `autogroup:self` or something similar would be useful without user creation - in a multi user tailnet it would avoid the need for one acl per user if you want to ensure use of named user accounts. It would be useful to have a toggle for whether the `@domain` part of the tailscale username is kept or removed, though this might play badly with host sharing if/when SSH is supported there.

I can see having users created just-in-time could be a useful option in some use cases but would probably need careful setup on the hosts to deal with things like sudo or setting the host password for the users. The two interplay if you're not happy with passwordless sudo.

On the other hand I would probably prefer to have ssh login refused if the user isn't already present on the host. If I'm managing users and groups another way (e.g. Ansible) I'd prefer to have a user login fail early so that the configuration issue can be spotted rather than end up with an inconsistently setup user.

> * If you want multiple individuals to log in to the same host user "devops", then what happens if there is already a session in process on the host for that user? Is another individual blocked from starting a session as that user? i.e. are the sessions exclusive?

I think this is a different use case as the non-dynamic-user part of this is already supported by the ACLs, though it comes back into the picture with just-in-time users with similar issues around groups and passwords.

I get the feeling that just-in-time user creation risks scope creep for tailscale - becoming more of an identification proxy rather than just reusing the authentication knowledge which tailscale already has.

**Comment by skorokithakis** [comment]
> If I'm managing users and groups another way (e.g. Ansible) I'd prefer to have a user login fail early so that the configuration issue can be spotted rather than end up with an inconsistently setup user.

Would you not transition to setting up the groups via Ansible, and then it would be fine to have users auto-created in the specific groups?

> I get the feeling that just-in-time user creation risks scope creep for tailscale - becoming more of an identification proxy rather than just reusing the authentication knowledge which tailscale already has.

Possibly, but I think the current situation of users being able to log in as anyone is inconvenient for most uses.

**Comment by gebrk** [comment]
> Possibly, but I think the current situation of users being able to log in as anyone is inconvenient for most uses.

Just to be clear - I agree with this statement, which is why I'm following this issue. But it appears to me that making `autogroup:self` work in the user acl but requiring the user to already exist on the target system would solve the issue for most use cases with far less development (and scope creep) than adding j-i-t user creation. 

> Would you not transition to setting up the groups via Ansible, and then it would be fine to have users auto-created in the specific groups?

Group membership isn't the entire picture, but if supplementary groups could be set per user that would help with that aspect. User password management is the other main practicality that would need to be sorted.

**Comment by skorokithakis** [comment]
That's true, I guess the need for password management makes JIT user creation much harder (a user that can sudo is pretty useless if they don't know their sudo password).

**Comment by jsharpe** [comment]
I'm running into the same requirement. The way I think this should be done is:
provide a `autogroup:ssh-username` field which does either a lookup of mapping tailscale users to ssh user accounts or uses a fixed mapping e.g. `alice@tailscale.com` maps to unix username `alice`. Note that ideally this would be done on a per tag basis as one group of hosts may have a different mapping scheme to another set of hosts (athough in my case we only have a single mapping scheme).

I don't need any JIT user creation - my servers use SSSD and pam to achieve this. I just need a way of restricting a tailscale user to a single username that follows a predefined mapping pattern i.e. `user@domain.com` -> `user`.

**Comment by nward** [comment]
My proposal from #7938 is something like the following - roughly...

```json
{
  "ssh_users": {
    "some_mapping_name": {
      "nward@domain": ["nward"],
      "foobar@domain": ["fbar"],
    },
    "another_mapping_name": {
      "nward@domain": ["nathan","wardn"],
      "foobar@domain": ["foo","barf"],
      "baz@domain": ["baz"],
    }
  },
  "ssh": [
    {
      "action": "accept",
      "src": ["group:group_a_users"],
      "dst": ["tag:group_a"],
      "users": ["ssh_users:some_mapping_name"],
    },
    {
      "action": "accept",
      "src": ["group:group_b_users"],
      "dst": ["tag:group_b"],
      "users": ["ssh_users:another_mapping_name"],
    }
  ]
}
```

**Comment by derek-burdick** [comment]
For enterprise use, the main configuration would be to include @domain or not. In addition, I typically see a few primary usernames:

shortname: ${firstinitial}${lastname} or some combination
namepart: ${first}.${last}
username: ${emailaddress}

Linux PAM/LDAP integrations use shortname, with just in time configuration of home directory and local user. SSSD on Redhat has an implementation, I have also used Azure AAD linux integration.
https://github.com/aad-for-linux/libnss-aad
https://github.com/aad-for-linux/aad-for-linux

**Comment by snowe2010** [comment]
It took me several hours of frustration until I learned that this _isn't_ how tailscale works. From the documents it was incredibly unclear to me that the users have to be already created. 

>With Tailscale, you can already connect machines in your network, and encrypt communications end-to-end from one point to another—and this includes, for example, SSHing from your work laptop to your work desktop. Tailscale also knows your identity, since that’s how you connected to your tailnet. When you enable Tailscale SSH, Tailscale claims port 22 for the Tailscale IP address (that is, only for traffic coming from your tailnet) on the devices for which you have enabled Tailscale SSH. This routes SSH traffic for the device from the Tailscale network to an SSH server run by Tailscale, instead of your standard SSH server. With Tailscale SSH, based on the ACLs in your tailnet, you can allow devices to connect over SSH and rely on Tailscale for authentication instead of public key authentication.

After rereading them about 6 times I finally was able to parse that it isn't what happens. I really was under the impression that Tailscale ssh would apply permissions to the users as well, based on the ACL. Either implementing this feature, or updating the docs to make it a _bit_ more clear that this isn't how it works would be really appreciated. Hopefully it can save some other people a few hours of their life.

**Comment by gdonval** [comment]
As I see it, Tailscale's marketing just misrepresents what Tailscale SSH really is: an off-band sysadmin tool.

E.g. when you would be tempted to run a SSH daemon to log onto prod/container once to check something specific on the running system, Tailscale SSH with "check" is perfect.

When one tool uses ssh to work (could just be rsync to create backups or some simple monitoring tool), then same thing but probably to a specific restricted user but with policy "accept".

Note that in both those cases, you will know what user you want to log on as.

The last main use case, that **seems** to be covered but is not because you create a split-head problem is what is asked for here. I would love that to work. I love that there is no "trust at first connection" with tailscale SSH but it's really not architected for it. It's gonna be hard to do it right because you need tailscale to basically implement PAM and all the mechanics found in OpenSSH.

**Comment by apagano-vue** [comment]
We have the same use-case as @nward: we need to map tailscale users to "Unix users" (e.g firstname.secondname@domain --> firstname). Currently there is a lot of duplication in our ACL configuration:

```
[...]
	"ssh": [
		{
			"action": "check",
			"src": ["user1@domain"],
			"dst": ["tag:ml"],
			"users": ["user1"],
		},

		{
			"action": "check",
			"src": ["user2@domain"],
			"dst": ["tag:ml"],
			"users": ["user2"],
		},
		{
			"action": "check",
			"src": ["user3@domain"],
			"dst": ["tag:ml"],
			"users": ["user3"],
		},

		[...]

	],
[...]
```

So the outlined solution proposed by @nward would be perfect for us as well.

**Comment by gdonval** [comment]
> we need to map tailscale users to "Unix users" 

What exactly is the problem with using normal OpenSSH for these? You get what you want without the split-head problem. 

What was proposed above can already be used, nearly as is, just by using OpenSSH and allowing or not access to port 22 on enpoints, just like you do with TailscaleSSH. If those endpoints have those users setup already, setting them up with a SSH certificate is really easy and solves the problem fully and cleanly, contrary to a mapping that you have to keep in sync with your AD/PAM config.

**Comment by apagano-vue** [comment]
> > we need to map tailscale users to "Unix users"
> 
> What exactly is the problem with using normal OpenSSH for these? You get what you want without the split-head problem.
> 
> What was proposed above can already be used, nearly as is, just by using OpenSSH and allowing or not access to port 22 on enpoints, just like you do with TailscaleSSH. If those endpoints have those users setup already, setting them up with a SSH certificate is really easy and solves the problem fully and cleanly, contrary to a mapping that you have to keep in sync with your AD/PAM config.

Forgive my ignorance/confusion but what would be the benefit of using Tailscale SSH then? Our understanding of Tailscale SSH is that it allows to simplify the management of SSH authentication. No more managing keys or certificates at scale, just outsource ssh auth to the tool you're already using to simplify your networking setup: Tailscale. So my understanding is that Tailscale SSH is effectively a web-based SSO flow for SSH authentication that is very easily enabled for all machines on your Tailnet. Ideally we'd like not to have to manage public keys or our own CA.

**Comment by snowe2010** [comment]
Yeah I have the same  understanding as @apagano-vue, but we must be missing something since this seems like it wasn't what TS SSH was designed for.

**Comment by nward** [comment]
I will second that @snowe2010 @apagano-vue.

We are using TS SSH because it handles authentication for us.

**Comment by gdonval** [comment]
Don't comment, just put thumbs up on the original request.

Regarding general puzzlement look at [this](https://github.com/tailscale/tailscale/issues/5562#issuecomment-1773723666).

---

> Our understanding of Tailscale SSH is that it allows to simplify the management of SSH authentication...

That is correct if what you need is to connect to a machine for administration or looking at the logs or to make services on two machines interact with each other through SSH. For the rest it's incorrect.

Typical use cases would be giving access to a `db_backup` user that can RO access a database on your `DatabaseNode`. Admin access to the database would be hidden behind Tailscale. You can give your `BackupServiceNode` access to `db_backup` on `DatabaseNode` to perform regular backups piped through SSH. Every time you redeploy `DatabaseNode` on one or more **different** node, its public address might change (or it might have none) but you will always have access through Tailscale without having to find a way to distribute SSH keys around.

Another, similar, use case would be backups with `rsync`. If what you wanna backup is your `nginx` web root, you might elect to use Tailscale to manage the routing between the nginx server and your backup servers. Each would be authorized to connect as `nginx` to only access what the server can access and proceed with the backup.

A different usecase is putting Tailscale in containers running single services. You probably don't want/need a full OpenSSH instance in there with full PAM/AD authentication when it's irrelevant. It's also not a normal usecase: even as yourself you probably don't want to have unrestricted access to something you **should redeploy** if something went wrong but you probably want a way to access it if absolutely necessary to troubleshoot. It's a container so you probably want to access `root` in `check` mode. But same thing again: you don't have to alter your containers to add openSSH, configure your unix users, add your keys, etc. Tailscale is all you need.

A different usecase: you manage your family computers. Sometimes people have problems in their session. You've put all those computers in a Tailscale network and you provided your user with `SSH` access to them. So you can `tailscale ssh dad@...` or  `tailscale ssh mom@...` or `tailscale ssh sister@...`, etc. for troubleshoot. Maybe even in `check` mode.

^ This BTW, is how Tailscale SSH works: it bypasses all authentication system on the computer when used in SSH mode. That's where the split-head problem comes from: your system's security is based on users permissions. Said permissions are provided from one source: the PAM system (possibly with remote backends like AD). Tailscale bypasses that entirely. Hell, by default it will let you log in as *anyone* as long as that user exists!

Now let's say that `sam`'s access was revoked in your AD/PAM. You forgot to update your user mapping in Tailscale? Well, `sam` still has the same access as before. You did not read the doc properly and assumed `sam` can only auth as themselves? Not only `sam` still has access but they have access as any user on the system.

----

Long story short, the current Tailscale SSH is devised around "automated"/service-based/troubleshooting-based kind of access without the trouble of deploying OpenSSH with a set of keys that are hard to rotate and update. 

You want to connect your DB with your web server but want your web server to be actually **spawned** by your DB server node to make absolutely sure your DB is up and running when the webserver becomes available? Easy-peasy! You put them node in a Tailscale network. You give the DB node SSH access to the web server node as the relevant (known in advance) user and you make it `systemctl start nginx` through SSH. Solved.

You need to troubleshoot that DB after something went wrong, you SSH with Tailscale onto that DB node, after SSO check as the main DB user, to troubleshoot what has gone wrong. Then you figure you need access to the logs so you check-ssh into the node as the log user to have a read. You fix the problem in your config and redeploy. 

Tailscale does not have the backend to sync its mapping with local access rules (PAM/AD) so it's a poor replacement for OpenSSH as used by human users (as much as I would love that to be implemented, it's a design catastrophe).

**Comment by snowe2010** [comment]
>That is correct if what you need is to connect to a machine for administration or looking at the logs or to make services on two machines interact with each other through SSH. For the rest it's incorrect.

@gdonval Do you work for TS? [Their documentation seems to completely contradict what you are saying.](https://tailscale.com/kb/1193/tailscale-ssh/#how-is-tailscale-ssh-different)

> Historically, to secure an SSH connection, you generate a keypair on the machine you are connecting from (known as the client), with the private key stored on the client, and the public key distributed to the device you want to connect to (known as the server). This lets the server authenticate communication from the client.
> 
> With Tailscale, you can already connect machines in your network, and encrypt communications end-to-end from one point to another—and this includes, for example, SSHing from your work laptop to your work desktop. Tailscale also knows your identity, since that’s how you connected to your tailnet. When you enable Tailscale SSH, Tailscale claims port 22 for the Tailscale IP address (that is, only for traffic coming from your tailnet) on the devices for which you have enabled Tailscale SSH. This routes SSH traffic for the device from the Tailscale network to an SSH server run by Tailscale, instead of your standard SSH server. With Tailscale SSH, based on the ACLs in your tailnet, you can allow devices to connect over SSH and rely on Tailscale for authentication instead of public key authentication.

**Comment by gdonval** [comment]
What I stated are facts of how it works now. Being able to connect as any user and bypassing PAM is *how it works*. Try it for yourself if you are unsure.

I was told by apagano that it if was the case, they don't see the point or how this can be branded as useful by Tailscale and that, just like yourself, he assumed it was a drop in replacement for OpenSSH but with the twist that key management becomes simpler.

It does make key management simpler but it's not a drop-in replacement. Nothing in the documentation indicates you can use existing auth machinery in Tailscale. Quite the opposite: most defaults rules they show either involve all users but root, or all users. Not really granular.

Most configuration examples simply show how a group (like "prod") can be made to access the production server both as root and non-root (i.e. as anyone) because that's the problem they solve **now**. Keep in mind that with those configs, `alice` (who is part of prod) can authenticate as `bob` on the prod server because that's what the rules say. In real servers though, there are probably no `alice` nor `bob` accounts. Probably just root and a few domain specific users (e.g. to manage the DB or do backups), whose names can always the same: that's what I detailed in my examples and that's super easy with Tailscale.

In your citation, they say you can connect from **your** work laptop to **your** work desktop. Good. That seems to imply it's just like OpenSSH, or does it? With the way *it currently works*, this implies that **you** own that tailscale net and that said computers are not multi-tenant (i.e. they were allocated to you and you alone). Or that your company created, syncs and maintains ACLs using their auth backend for you (which **is** possible with current syntax and will require ACL generation, hence why this report won't receive attention). One mistake though and you're root or you can impersonate colleagues. Not nice. And even if well-synced, changes on one side won't be propagated instantly. 

With OpenSSH you have the burden of distributing keys. With Tailscale you have the burden of maintaining error-prones ACLs. It's trading a burden for another. The day Tailscale gains the ability to talk to PAM and source certificates, it will become an improved drop-in replacement for OpenSSH. But until then, at work, its SSH part only really shines at the level of "my production team must be given root access on production servers" or "our DB tool needs to be given access to the DBs to do backups". In dynamic environments (like typically on the cloud), tailscale solves all of that "hardcoded" stuff extrememy well, without having to think about key management! To connect your desktop to your laptop though, especially if multitenant scenarios with a big number of users, you're much better off dumping your OpenSSH keys there and manage access to physical machines through ACL (if within the confines of a Tailscale network).

**Comment by danchitnis** [comment]
So, based on my understanding from the discussions above, Tailscale's SSH is *intended* to connect any user to any user. On the other hand, the business model of Tailscale is to get paid by the number of users on larger teams. So I fail to understand how this Tailsclae's SSH could work for larger teams. When I have 10 users in my teams, and some of them are sudo, and some aren't, I do not want every user to log into each other and create a security risk and lose their privacy. My intended use of Tailscale is as a VPN to access compute machines and servers. This idea may work for a small team (which is ironically free), but not for a large team which is paid!

**Comment by gdonval** [comment]
Look, let's put it this way: Tailscale **SSH** is, currently, a management and debugging tool that you manage with ACls. That's all. 

It's not a replacement for OpenSSH, it does not have what's required to do that. It's not because the SSH protocol is used that the experience will be the same as OpenSSH's. It is evidently not the case.

**Comment by chrisckchang** [comment]
New as of May 7 2024: 
`localpart:*@<domain>` ACL autogroup allows Tailscale SSH access to a user on the host whose name matches the local-part of the user’s Tailscale login

More details here: https://tailscale.com/kb/1193/tailscale-ssh#users and under "SSH access rule examples"
